src/os/bluestore/Allocator.h

```diff
@@ -16,8 +16,207 @@
 #include <ostream>
 #include "include/ceph_assert.h"
 #include "bluestore_types.h"
+#include "common/ceph_mutex.h"
+
+typedef interval_set<uint64_t> release_set_t;
+typedef release_set_t::value_type release_set_entry_t;
 
 class Allocator {
+protected:
+
+  struct ExtentCollectionTraits {
+    size_t num_buckets;
+    size_t base_bits; // min extent size
+    size_t base = 1ull << base_bits;
+    size_t factor;  // single bucket size range to be
```

Comment:
In ExtentCollectionTraits there are two facets: linear and geometric.
Those cannot be used intermixed.
One must be linear or geometric from creation up until destruction.
Maybe it should be split into `LinearBucket[Trait]` and `ExpBucket[Trait]` ?

Reply:
Done